pkgs/by-name/xa/xar/package.nix

```diff
@@ -0,0 +1,178 @@
+{
+  lib,
+  stdenv,
+  fetchurl,
+  applyPatches,
+
+  autoreconfHook,
+
+  # Required dependencies.
+  zlib,
+  libxml2,
+
+  # Optional dependencies.
+  withExt2 ? lib.meta.availableOn stdenv.hostPlatform e2fsprogs,
+  e2fsprogs,
+  withAcl ? lib.meta.availableOn stdenv.hostPlatform acl,
+  acl,
+  withBzip2 ? true,
+  bzip2,
+  withLzma ? true,
+  xz, # lzma
+
+  # Platform-specific dependencies.
+  musl-fts,
+  openssl,
+  darwin, # CommonCrypto
+
+  gitUpdater,
+
+  # for tests
+  xar,
+  testers,
+  python3,
+  libxslt, # xsltproc
+  runCommand,
+  makeWrapper,
+  buildPackages,
+}:
+let
+  pname = "xar";
+  version = "498";
+
+  pythonForTests = py: py.withPackages (p: [ p.xattr ]);
+
+  src = applyPatches {
+    name = "source";
+    src = fetchurl {
+      url = "https://opensource.apple.com/tarballs/xar/xar-${version}.tar.gz";
+      hash = "sha256-nO5PgLls9ZLMxUWk/dUeTaSlvTtHNJAWN9Z7BD7/PHU=";
+    };
+    patches = lib.filesystem.listFilesRecursive ./patches;
+  };
+in
+stdenv.mkDerivation {
+  inherit pname version src;
+
+  sourceRoot = "source/xar";
+
+  outputs = [
+    "out"
+    "lib"
+    "dev"
+  ];
+
+  separateDebugInfo = true;
+
+  strictDeps = true;
+
+  nativeBuildInputs = [ autoreconfHook ];
+
+  # For some reason libxml2 package headers are in subdirectory and thus aren’t
+  # picked up by stdenv’s C compiler wrapper (see ccWrapper_addCVars). This
+  # doesn’t really belong here and either should be part of libxml2 package or
+  # libxml2 in Nixpkgs can just fix their header paths.
+  env.NIX_CFLAGS_COMPILE = "-isystem ${libxml2.dev}/include/libxml2";
+
+  buildInputs =
+    let
+      crypto = if stdenv.hostPlatform.isDarwin then darwin.CommonCrypto else openssl;
```

Comment:
If there would be more linux/darwin specific packages it would probably be better to add them like the musl one

Reply:
Hm, are you suggesting to use `lib.optional` in this case? 🤔

I’m not really sure about this, `musl-fts` is an additional platform-specific dependency, while OpenSSL and CommonCrypto is a platform-specific choice between different cryptography libraries that I feel like is better expressed with `if … then … else` compare to two distinct `lib.optional`s. That said, I don’t have a strong preference for either.

Reply:
In a matter of months I believe `CommonCrypto` will be available without any dependency on macOS as part of the SDK rework, so any colour of bikeshed is fine here.

Reply:
Would be nice to use `CommonCrypto` for non-Darwin platforms as well since it looks like OpenSSL support was abandoned by Apple. Gentoo does the opposite and always uses OpenSSL even on Darwin.
https://gitweb.gentoo.org/repo/gentoo.git/tree/app-arch/xar/xar-1.8.0.0.498.ebuild?id=f09efc0fdb5cfa3271f628d0d8508e786fecab50#n57

Reply:
If OpenSSL works on Darwin, I see no reason not to use it on Darwin. The Darwin stdenv bootstrap already requires OpenSSL due to ld64 (if not other packages, but I know ld64 does for sure because I wrote it that way).

Reply:
Seems reasonable to me if we’re having to maintain OpenSSL support for Linux anyway. Is using `CommonCrypto` on Linux at all viable, though, if that lets us defer maintenance to Apple?

Reply:
CommonCrypto depends on corecrypto, which is extremely non-free.

Reply:
As far as I can tell, xar is using CommonCrypto for hashing. It’s the same situation as ld64, but the work has already been done to support using OpenSSL for hashing (rather than having to write compatibility shims like I did).

Reply:
Perhaps let’s use OpenSSL on every platform, then, if we’re having to maintain that support anyway and `CommonCrypto` isn’t a option. Not really a big deal either way though.

Reply:
Done.